format/fit/fit.go

```diff
@@ -0,0 +1,337 @@
+
+}
+
+func fieldUint(fieldFn func(string, ...scalar.UintMapper) uint64, expectedSize uint64, fDef mappers.FieldDef, uintFormatter scalar.UintFn, fdc *fileDescriptionContext) {
+	var val uint64
+
+	if fDef.Size != expectedSize {
+		arrayCount := fDef.Size / expectedSize
+		for i := uint64(0); i < arrayCount; i++ {
+			fieldFn(fmt.Sprintf("%s_%d", fDef.Name, i), uintFormatter)
+		}
+	} else {
+		if uintFormatter != nil {
+			val = fieldFn(fDef.Name, uintFormatter)
+		} else {
+			val = fieldFn(fDef.Name)
+		}
+
+		// Save developer field definitions
+		switch fDef.Name {
+		case "developer_data_index":
+			fdc.devIdx = val
+		case "field_definition_number":
+			fdc.fDefNo = val
+		case "fit_base_type_id":
+			fdc.typ = mappers.TypeDefMap["fit_base_type"][val].Name
+		case "native_field_num":
+			fdc.nativeFieldNo = val
+		case "native_mesg_num":
+			fdc.nativeMsgNo = val
+		}
+	}
+}
+
+func fieldSint(fieldFn func(string, ...scalar.SintMapper) int64, expectedSize uint64, fDef mappers.FieldDef, sintFormatter scalar.SintFn) {
+	if fDef.Size != expectedSize {
+		arrayCount := fDef.Size / expectedSize
+		for i := uint64(0); i < arrayCount; i++ {
+			fieldFn(fmt.Sprintf("%s_%d", fDef.Name, i), sintFormatter)
+		}
+	} else {
+		if sintFormatter != nil {
+			fieldFn(fDef.Name, sintFormatter)
+		} else {
+			fieldFn(fDef.Name)
+		}
+	}
+}
+
+func fieldFloat(fieldFn func(string, ...scalar.FltMapper) float64, expectedSize uint64, fDef mappers.FieldDef, floatFormatter scalar.FltFn) {
+	if fDef.Size != expectedSize {
+		arrayCount := fDef.Size / expectedSize
+		for i := uint64(0); i < arrayCount; i++ {
+			fieldFn(fmt.Sprintf("%s_%d", fDef.Name, i), floatFormatter)
+		}
+	} else {
+		fieldFn(fDef.Name)
+	}
+}
+
+func fieldString(d *decode.D, fDef mappers.FieldDef, fdc *fileDescriptionContext) {
+	val := d.FieldUTF8NullFixedLen(fDef.Name, int(fDef.Size), scalar.StrMapDescription{"": "Invalid"})
+
+	// Save developer field definitions
+	switch fDef.Name {
+	case "field_name":
+		fdc.name = val
+	case "units":
+		fdc.unit = val
+	}
+}
+
+func fitDecodeDataMessage(d *decode.D, drc *dataRecordContext, lmfd localFieldDefMap, dmfd devFieldDefMap, isDevMap localMsgIsDevDef) {
+	var fdc fileDescriptionContext
+	keys := make([]int, len(lmfd[drc.localMessageType]))
+	i := 0
+	for k := range lmfd[drc.localMessageType] {
+		keys[i] = int(k)
+		i++
+	}
+	sort.Ints(keys)
+
+	isDevDep := isDevMap[drc.localMessageType]
+
+	for _, k := range keys {
+		fDef := lmfd[drc.localMessageType][uint64(k)]
+
+		var uintFormatter = mappers.GetUintFormatter(fDef)
+		var sintFormatter = mappers.GetSintFormatter(fDef)
+		var floatFormatter = mappers.GetFloatFormatter(fDef)
+
+		switch fDef.Type {
+		case "enum", "uint8", "uint8z", "byte":
+			fieldUint(d.FieldU8, 1, fDef, uintFormatter, &fdc)
+		case "uint16", "uint16z":
+			fieldUint(d.FieldU16, 2, fDef, uintFormatter, &fdc)
+		case "uint32", "uint32z":
+			fieldUint(d.FieldU32, 4, fDef, uintFormatter, &fdc)
+		case "uint64", "uint64z":
```

Comment:
It's already handled by the uintFormatter in mappers/values.go.

Reply:
I think I've figured out the *z variants. They are used as bitflags, meaning 0 is invalid because then it lacks any value. It also means another weekend project for me =)

Reply:
> It's already handled by the uintFormatter in mappers/values.go.

Aha yes forgot that

> I think I've figured out the *z variants. They are used as bitflags, meaning 0 is invalid because then it lacks any value. It also means another weekend project for me =)

:) that makes sense, and the flags are defined somewhere?

But let me know if want to do it as separate future PR or so

Reply:
There aren't that many of them, but they are defined in the Profile.xlsx.
None of their example files nor mine have any of those fields though, so I can't test it either way. I'm good for merging now if you are. Do you want me to squash rebase it first?

Reply:
Aha i see, add if you happen to find files using them.

In this case i would probably squash into one or very few commits, but the commit history looks reasonable so i'm ok to merge as is also.